File: apps/xunit-example/WrappedStepsStoryTest.cs
using ExecutableStories.Xunit;
using Xunit;
using XunitExample;

namespace XunitExample.Tests;

public class WrappedStepsStoryTest
{
    [Fact]
    public void Calculator_adds_using_Fn_and_Expect()
    {
        Story.Init("Calculator adds two numbers using Fn and Expect");

        var a = Story.Fn("Given", "number a is 5", () => 5);
        var b = Story.Fn("Given", "number b is 3", () => 3);

        var result = Story.Fn("When", "the numbers are added", () => Calculator.Add(a, b));

        Story.Expect("the result is 8", () => Assert.Equal(8, result));

        Story.RecordAndClear();
    }

    [Fact]
    public void Calculator_subtracts_using_Fn_with_timing()
    {
        Story.Init("Calculator subtracts using Fn with timing");

        Story.Fn("Given", "two numbers 10 and 4", () => { });

        var result = Story.Fn("When", "the second is subtracted from the first",
            () => Calculator.Subtract(10, 4));

        Story.Expect("the result is 6", () => Assert.Equal(6, result));

        Story.RecordAndClear();
    }

    [Fact]
    public void Calculator_division_by_zero_captured_in_Fn()
    {
        Story.Init("Calculator division by zero captured in Fn");

        Story.Fn("Given", "a number 10 and zero", () => { });

        Story.Expect("division by zero throws an error", () =>
            Assert.Throws<DivideByZeroException>(() => Calculator.Divide(10, 0)));

        Story.RecordAndClear();
    }

    [Fact]
    public void Mixed_markers_and_wrapped_steps()
    {
        Story.Init("Mixed markers and wrapped steps");

        Story.Given("the calculator is ready");

        var result = Story.Fn("When", "we multiply 7 by 6", () => Calculator.Multiply(7, 6));

        Story.Expect("the result is 42", () => Assert.Equal(42, result));

        Story.And("the result is a positive number");
        Assert.True(result > 0);

        Story.RecordAndClear();
    }
}

File: apps/junit5-example/src/test/kotlin/example/WrappedStepsStoryTest.kt
package example

import dev.executablestories.junit5.Story
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import java.util.function.Supplier

class WrappedStepsStoryTest {

    @Test
    fun calculatorAddsUsingFnAndExpect() {
        Story.init("Calculator adds two numbers using fn and expect")

        val a = Story.fn("Given", "number a is 5", Supplier { 5 })
        val b = Story.fn("Given", "number b is 3", Supplier { 3 })

        val result = Story.fn("When", "the numbers are added", Supplier { Calculator.add(a, b) })

        Story.expect("the result is 8", Runnable { assertEquals(8, result) })
    }

    @Test
    fun calculatorSubtractsUsingFnWithTiming() {
        Story.init("Calculator subtracts using fn with timing")

        Story.fn("Given", "two numbers 10 and 4", Runnable {})

        val result = Story.fn("When", "the second is subtracted from the first",
            Supplier { Calculator.subtract(10, 4) })

        Story.expect("the result is 6", Runnable { assertEquals(6, result) })
    }

    @Test
    fun calculatorDivisionByZeroCapturedInFn() {
        Story.init("Calculator division by zero captured in fn")

        Story.fn("Given", "a number 10 and zero", Runnable {})

        Story.expect("division by zero throws an error",
            Runnable { assertThrows(ArithmeticException::class.java) { Calculator.divide(10, 0) } })
    }

    @Test
    fun mixedMarkersAndWrappedSteps() {
        Story.init("Mixed markers and wrapped steps")

        Story.given("the calculator is ready")

        val result = Story.fn("When", "we multiply 7 by 6", Supplier { Calculator.multiply(7, 6) })

        Story.expect("the result is 42", Runnable { assertEquals(42, result) })

        Story.and("the result is a positive number")
        assert(result > 0)
    }
}
